### ISSUE TITLE
ChannelsLiveServerTestCase ReactorNotRestartable exception

### ISSUE THREAD

**Issue by brianmay** [django/channels#897]
Yes, I updated all channels related stuff this morning :-)

```
Traceback (most recent call last):
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/nose/suite.py", line 316, in setupContext
    try_run(context, names)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/django/test/testcases.py", line 1352, in setUpClass
    raise cls.server_thread.error
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/channels/testing/live.py", line 28, in run
    self.daphne.run()
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/server.py", line 132, in run
    reactor.run(installSignalHandlers=self.signal_handlers)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 266, in run
    self.startRunning(installSignalHandlers=installSignalHandlers)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/twisted/internet/base.py", line 1222, in startRunning
    ReactorBase.startRunning(self)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/twisted/internet/base.py", line 730, in startRunning
    raise error.ReactorNotRestartable()
twisted.internet.error.ReactorNotRestartable: 
```

**Comment by andrewgodwin** [comment]
Can I get a bit more context than this? What sort of tests are you trying to run? Is there a `LiveTestCase` variant involved here?

**Comment by brianmay** [comment]
Tried to create a simpler test case, but so far without success :-(.

We are running tests using nose,  harvest, and aloe. The ``TestCase`` is defined as:

```
class TestCase(ChannelsLiveServerTestCase, AloeTestCase):
    pass
```

Oh, wonder if I am looking at a symptom of the error, not the cause. The above error seems to occur constantly immediately after I get this error:

```
Traceback (most recent call last):
  File "/home/brian/.pyenv/versions/3.6.4/lib/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 290, in run
    f(*args, **kwargs)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/server.py", line 274, in timeout_checker
    protocol.check_timeouts()
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/http_protocol.py", line 243, in check_timeouts
    self.basic_error(503, b"Service Unavailable", "Application failed to respond within time limit.")
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/http_protocol.py", line 276, in basic_error
    (b"Content-Type", b"text/html; charset=utf-8"),
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/http_protocol.py", line 196, in handle_reply
    raise ValueError("HTTP response has already been started")
```

**Comment by andrewgodwin** [comment]
OK, `ChannelsLiveServerTestCase` explains the restarting thing, as it's the only component that is going to worry about the reactor. I have reproduced it here with this code:

```class SomeLiveTests(ChannelsLiveServerTestCase):

    def test_live_stuff(self):
        print(self.live_server_url)

class SomeOtherLiveTests(ChannelsLiveServerTestCase):

    def test_live_stuff2(self):
        print(self.live_server_url)
```

**Comment by brianmay** [comment]
Ok, yes, that is a much simpler test case, thanks. (think you missed the first line...)

I can reproduce this on demand now.

So, looks like it only fails on the 2nd test case, this might explain some of the randomnesses I was getting.

Probably the "HTTP response has already been started" is unrelated.

**Comment by andrewgodwin** [comment]
Indeed, or that is a symptom of the underlying problem (that we're trying to start a Twisted reactor twice and twisted REALLY DOES NOT LIKE THIS). I'm sure I can convince it to restart.

**Comment by brianmay** [comment]
Oh, I guess we are attempting to restart Daphne between tests, which restarts the Twisted reactor. Ok, this makes sense now...

**Comment by brianmay** [comment]
I have found two kludges that seem to work:

1. Modify Daphne not to shut down the reactor each time, and if the reactor is still running at start, do not to restart it.
2. Modify Daphne to set `reactor._startedBefore=False` after shutting down the reactor.

However, in both cases, this breaks the `self.join()` call in `DaphneServerThread.terminate()`` which hangs forever. I understand why for the first case, not sure I understand for the 2nd case. So I commented it out this call.

So, yes, somewhat a kludge, but it works :-)

**Comment by andrewgodwin** [comment]
I'm not sure those kludges are going to be something I can put in an actual release, but thanks, they might help debug.

**Comment by brianmay** [comment]
Correction: 2nd kludge not working so well. Need to set `reactor._started = False` and `reactor.running = False` too, but then things get confused with errors like "RuntimeError: This event loop is already running". Will try 1st kludge again.

**Comment by brianmay** [comment]
Still getting the "HTTP response has already been started." seems to be somewhat random in nature. Will try to track it down. I don't think this is due to my kludge...

**Comment by brianmay** [comment]
Oh, the real error is "Application failed to respond within time limit." The error handling seems to turn this into "HTTP response has already been started.", which is kind of confusing. This could be because  I am not shutting down things correctly between tests, especially after this kludge.

```
Traceback (most recent call last):
  File "/home/brian/.pyenv/versions/3.6.4/lib/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 290, in run
    f(*args, **kwargs)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/server.py", line 279, in timeout_checker
    protocol.check_timeouts()
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/http_protocol.py", line 243, in check_timeouts
    self.basic_error(503, b"Service Unavailable", "Application failed to respond within time limit.")
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/http_protocol.py", line 276, in basic_error
    (b"Content-Type", b"text/html; charset=utf-8"),
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/http_protocol.py", line 196, in handle_reply
    raise ValueError("HTTP response has already been started")
ValueError: HTTP response has already been started
```

**Comment by brianmay** [comment]
My feeling at this stage is that we need to fork Daphne to another process. So we can kill it and the reactor with it.

I have implemented such a solution. It seemed to work initially, but now it causes problems with selenium. Still investigating.

**Comment by andrewgodwin** [comment]
We had to do that in Channels 1, it was a massive pain, I was hoping to avoid it again.

If you are considering it, though, the Daphne test suite already has to do this, so maybe we can promote that into Daphne itself:  https://github.com/django/daphne/blob/master/tests/http_base.py#L13

**Comment by brianmay** [comment]
Ok, current status is:

* Server works perfectly running in a separate process.
* Try to navigate to the page with selenium works, but then the browser closes the socket prematurely without a response (confirmed with wireshark). This gives ``http.client.RemoteDisconnected: Remote end closed connection without response``.
* Identical results with both latest versions of Firefox and Chrome.
* If I tell selenium to connect instead to a process that I started manually outside the test process, it also works.

**Comment by proofit404** [comment]
I briefly look at live server test case.

It uses threads to run daphne in parallel with tests. That means it should start and stop reactor between test. I know we use asyncio event loop, which supports start and stop procedures on the created loop. Does it mean twisted can do the same if it uses asyncio reactor?

**Comment by andrewgodwin** [comment]
Unfortunately the Twisted asyncio reactor is not quite as clean as the asyncio event loop and can't be restarted (there's enough global state in Twisted that it's hard). I can ping the Twisted people I know about this, but the answer is likely not going to be "oh it's easy to restart the reactor!"

Given I didn't want to expend the energy to rewrite daphne into pure asyncio and aiohttp as part of Channels 2 (there was enough work!) keeping Twisted and plugging in the async reactor made the most sense, but it does end up with weird situations like this.

I'm pretty sure you can fake Twisted into restarting a new reactor by deleting `twisted.internet.reactor` from `sys.modules` and re-running `asyncioreactor.install()`, as that's the trick we use already to uninstall non-async reactors if they're discovered: https://github.com/django/daphne/blob/master/daphne/server.py#L14 - @brianmay , did you try that?

**Comment by proofit404** [comment]
The reason I ask about that is that we used multiprocessing in Channels 1. If there are two or more live server test cases, the second one will encounter reactor start error in the test setup code. Does multiprocessing stuff was removed intentionally and we don't need this stuff already. Or running live server test case multiple times is the problem with Channels 2?

**Comment by brianmay** [comment]
No, I haven't tried deleting `twisted.internet.reactor` - will try that tomorrow.

**Comment by andrewgodwin** [comment]
@proofit404 The main reason for multiprocessing in Channels 1 was because we couldn't easily mix the worker and server processes. Now they're all in the same process, I had hoped to remove that restriction, and I still think we might be able to get round the reactor.

**Comment by proofit404** [comment]
As far as I remember I wrap daphne and worker into multiprocessing instead of two separate threads exactly because of reactor restart error.

**Comment by andrewgodwin** [comment]
Ugh, right, I must have remembered wrong. Well, we have that as a fallback if we need it.

**Comment by brianmay** [comment]
The following change:

```diff
diff --git a/channels/testing/live.py b/channels/testing/live.py
index 92264e2..1651610 100644
--- a/channels/testing/live.py
+++ b/channels/testing/live.py
@@ -1,3 +1,5 @@
+import sys
+
 from django.conf import settings
 from django.db import connections
 from django.test.testcases import LiveServerTestCase, LiveServerThread
@@ -63,6 +65,8 @@ class DaphneServerThread(LiveServerThread):
         if hasattr(self, "daphne"):
             # Stop the WSGI server
             self.daphne.stop()
+            if 'twisted.internet.reactor' in sys.modules:
+                del sys.modules['twisted.internet.reactor']
         self.join()
```

Still gives ``twisted.internet.error.ReactorNotRestartable``.

My theory is the reactor is not getting deleted because Daphne is still holding a reference.

**Comment by brianmay** [comment]
Not particularly elegant, but the simple test case now works:

```
diff --git a/channels/testing/live.py b/channels/testing/live.py
index 92264e2..e15326c 100644
--- a/channels/testing/live.py
+++ b/channels/testing/live.py
@@ -1,3 +1,6 @@
+import importlib
+import sys
+
 from django.conf import settings
 from django.db import connections
 from django.test.testcases import LiveServerTestCase, LiveServerThread
@@ -63,6 +66,15 @@ class DaphneServerThread(LiveServerThread):
         if hasattr(self, "daphne"):
             # Stop the WSGI server
             self.daphne.stop()
+
+            if 'twisted.internet.reactor' in sys.modules:
+                del sys.modules['twisted.internet.reactor']
+                self.daphne = None
+                del sys.modules['daphne.server']
+                server = importlib.import_module('daphne.server')
+                global Server
+                Server = server.Server
+
         self.join()
```

Note the conditional not strictly required, but sometimes helps with the debugging. If the ``twisted.internet.error.ReactorNotRestartable`` error is triggered, the tests will try to clean up, and exceptions at this stage will hide the original error.

**Comment by brianmay** [comment]
Still randomly getting [the error I previously reported](https://github.com/django/channels/issues/897#issuecomment-364806829), I suspect sometimes the consumer is being shutdown at end of test while daphne is still waiting for response.

**Comment by brianmay** [comment]
The good news is my tests seem to be working now. Most of the time anyway...

**Comment by brianmay** [comment]
Also seeing this error randomly. I suspect might be some sort of race condition shutting things down at the end of each test.

```
[2018-02-14 09:53:16,077: ERROR/daphne.server] Exception inside application: 'NoneType' object has no attribute 'writeHeaders'
  File "/home/brian/tree/myro/channels/channels/http.py", line 191, in __call__
    await self.handle(body)" with "Hyp3S3cUR3P455"
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/asgiref/sync.py", line 95, in __call__ssword"
    return await asyncio.wait_for(future, timeout=None)3S3cUR3P455"
  File "/home/brian/.pyenv/versions/3.6.4/lib/python3.6/asyncio/tasks.py", line 339, in wait_for
    return (yield from fut)
  File "/home/brian/.pyenv/versions/3.6.4/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/asgiref/sync.py", line 110, in thread_handler
    return self.func(*args, **kwargs)
  File "/home/brian/tree/myro/channels/channels/http.py", line 227, in handle
    self.send(response_message)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/asgiref/sync.py", line 49, in __call__
    return call_result.result()
  File "/home/brian/.pyenv/versions/3.6.4/lib/python3.6/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/home/brian/.pyenv/versions/3.6.4/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/asgiref/sync.py", line 63, in main_wrap
    result = await self.awaitable(*args, **kwargs)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/server.py", line 196, in handle_reply
    protocol.handle_reply(message)
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/daphne/http_protocol.py", line 210, in handle_reply
    http.Request.write(self, message.get("body", b""))
  File "/home/brian/.pyenv/versions/3.6.4/envs/myrocc/lib/python3.6/site-packages/twisted/web/http.py", line 1069, in write
    self.channel.writeHeaders(version, code, reason, headers)
  'NoneType' object has no attribute 'writeHeaders'
```

**Comment by andrewgodwin** [comment]
Yeah, I don't see a path this leads down that doesn't end up with it being a really nasty hack. We likely have to go back to separate processes.

**Comment by brianmay** [comment]
Created another issue for that last stack trace, I think this was unrelated. #904.

**Comment by brianmay** [comment]
For the record, another approach involved something similar to the following code, which IMHO, is worse:

```diff
+                del sys.modules['twisted.internet.reactor']
+                asyncioreactor.install()
+                reactor = importlib.import_module('twisted.internet.reactor')
+                assert reactor is not None
+                daphne.server.reactor = reactor
+                daphne.server.current_reactor = reactor
```

**Comment by brianmay** [comment]
Latest single-process hack: https://github.com/brianmay/channels/commit/9345cdc0c81861742d93c275471e9dab283867e6

Latest multi-process hack:
https://github.com/brianmay/channels/commit/842d6758ad73b6a21ed052a3e3ec0cd93f6b6377

Currently, the multi-process hack breaks selenium - wish I know why. This is my first attempt at using multiprocessing, so it is possible I have got something wrong.

**Comment by brianmay** [comment]
Oh, somehow my port number is getting corrupted in the multiprocessing version. So the ``run()`` function sets ``self._port`` correctly, but the ``port()`` property function returns 0. It looks like the Django test case is overwriting my value with 0. Might make some changes here.

**Comment by brianmay** [comment]
False hope, accidentally left debugging statement in that was breaking things.

**Comment by brianmay** [comment]
(I deleted some comments that were wrong and potentially confusing)

**Comment by andrewgodwin** [comment]
Any solution that involves separate processes I would like to share code with the Daphne test suite that launches separate processes to test request/response handling, ideally. I'm afraid I don't know what might or might not break Selenium, you'll have to look into that - I regard a test server as "working" when it is correctly serving HTTP requests and tells you where they are served from :)

**Comment by brianmay** [comment]
How is the expires header set (by default)? I notice that I get very different results for the two cases:

single process: Expires: Fri, 28 Apr 2017 19:40:00 GMT
multiple processes: Expires: Wed, 14 Feb 2018 03:28:19 GMT

Not quite comprehending why one has the current time, and the other has last year. Makes me wonder if the multiple process daphne is missing settings or something.

**Comment by brianmay** [comment]
Expires being different was a red herring. The test was using the Django logout page that is using the never_cache decorator, which sets the Expires header based on the current time. The tests use freezegun to fake the current time. Freezegun obviously isn't going to work when the process is running in a separate thread. Something we will need to think about later.

More related to the selenium issue. My recent tests show that Daphne takes 10 seconds to come up with a response. Curiously exactly 10 seconds two times in a row. I can't help but think I broke something in my multiprocessing fork.

**Comment by andrewgodwin** [comment]
This is one of the problems with multiprocessing, it's a mini-nightmare to make everything work correctly across processes, and sort of why I wanted to avoid it. I don't have great answers right now though - I've expended most of my mental energy fixing the nasty Redis event loop weirdness of #859.

**Comment by brianmay** [comment]
Thanks. Some of the messages I was seeing also looked very much like #859 too.

**Comment by brianmay** [comment]
Through liberal use of prints scattered throughout the code, I have found the 10 seconds delay happens when Django calls ``importlib.import_module`` in ``django.urls.resolvers`` function ``RegexURLResolver.urlconf_module`` in order to load the urls.

Can't help but think this is going to be some sort of weird locking issue. Maybe locks are being shared between the processes when they shouldn't be or something. I do see [this lock](https://docs.python.org/3.6/library/imp.html#imp.lock_held) for importing modules, but ``imp.lock_held()`` returns False.

**Comment by brianmay** [comment]
I am going to look at how Channels 1.x did this, see if that offers any insights.

**Comment by brianmay** [comment]
Not sure what is different, but that is looking a lot better now...

**Comment by brianmay** [comment]
This has a mixture of code from 1.x and 2.x, so far it seems to be working properly with selenium.

https://github.com/django/channels/compare/master...brianmay:multi_process_hack_v2

**Comment by andrewgodwin** [comment]
It would be nice if we could work out what part of that mixture makes the ten-second lag go away. At this point it looks like it would probably be beneficial to put most of this logic into Daphne itself for its own tests and then add a few Django-specific bits into the Channels subclass.

**Comment by brianmay** [comment]
Unfortunately my time may be restricted now. I can't see any significant differences between the two, just minor differences such as:

* The 2nd version subclasses ``multiprocessing.Process``.
* The 1st version uses ``multiprocessing.Queues`` to pass back the port, while the 1st version uses ``multiprocessing.Value``.
* In the 2nd version I fixed a bug where we could call ``self.is_ready.set()`` potentially before we have received the valid port.
* The 2nd version has stuff concerning overriding or modifying settings.

I guess I should dump the settings, see if there is any modifications.

**Comment by brianmay** [comment]
The second version also calls ``django.setup(set_prefix=False)``, and explicitly does some sort of database configuration.

**Comment by andrewgodwin** [comment]
Ah, the `setup` call would be my suspicion. I'll take a look at this myself over the weekend hopefully.

**Comment by andrewgodwin** [comment]
OK, I have rearranged the Daphne testing code to use `multiprocessing` and arranged it as a class that should be reuseable for the live test case: https://github.com/django/daphne/commit/853771ec956123c33ce8b3499c5fe4543c248295

**Comment by brianmay** [comment]
I updated the Channels stuff. Works, but some minor changes I suspect may be required. I suspect the in memory check is no longer required for example.

https://github.com/django/channels/compare/master...brianmay:multi_process_v3

**Comment by brianmay** [comment]
Looks like the Django logging configuration is different in the subprocess. I still need to determine if this is a problem with my code or not. I suspect that the subprocess may not be getting my Django logging config.

I also note that some of my tests that weren't failing before are now failing. Suspect I will find this is due to problems with my own code.

Apart from that, my tests seem to be working fine now.

**Comment by brianmay** [comment]
I think what was happening with logging for me was:

* channels 1: Django configured logging, which [nose](http://nose.readthedocs.io/en/latest/plugins/logcapture.html) changed. When we forked the worker process, we called `Django.setup()`, which replaced the logging with the pristine Django config.

* channels 2: Django configured logging, which nose overwrote. Which gives the Daphne process the nose modified values.

I tend to think Django channels 2 gets it right here.

It is perhaps worth noting that the nose logcapture, while probably a good idea, won't work when messages are being logged in another process. So, for now, I have disabled the nose logcapture.
